### PR TITLE
Bi-fidelity active learning for forward UQ

### LINKED ISSUE #23315: Bi-fidelity active learning for forward UQ
## Reason
Use Gaussian process active learning to selectively call a high-fidelity model during UQ while by default relying on a physics-based low-fidelity model.

## Design
Associated changes made to the `ActiveLearningGPDecision.C` class in terms of the learning functions.

## Impact
No expected impact to existing objects.

### PR DESCRIPTION
**Pull request by somu15**
closes #23315

### CODE REVIEW

**Comment by somu15** [comment]
@zachmprince @grmnptr a gentle reminder :)

File: modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h [hunk 1]
```diff
@@ -76,6 +76,11 @@ class ActiveLearningGPDecision : public ActiveLearningReporterTempl<Real>,
   /// The learning function parameter
   const Real & _learning_function_parameter;
 
+  /// Store all the input vectors used for training
+  std::vector<std::vector<Real>> _inputs_batch;
+  /// Store all the outputs used for training
+  std::vector<Real> _outputs_batch;
+
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h:83]: Why did you need to move these?

File: modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h [hunk 2]
```diff
@@ -39,32 +39,32 @@ class ActiveLearningGPDecision : public ActiveLearningReporterTempl<Real>,
...
    */
-  void setupData(const std::vector<std::vector<Real>> & inputs, const std::vector<Real> & outputs);
+  virtual void setupData(const std::vector<std::vector<Real>> & inputs,
+                         const std::vector<Real> & outputs);
 
   /**
-   * This makes decisions whether to call the full model or not based on
-   * GP prediction and uncertainty.
+   * This evaluates the active learning acquisition function and returns bool
+   * that indicates whether the GP model failed.
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h:61]:
```suggestion
   * This method evaluates the active learning acquisition function and returns bool
   * that indicates whether the GP model failed.
```

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+   * This is where most of the computations happen:
+   *   - Data is accumulated for training
+   *   - GP models are trained
+   *   - Decision is made whether more data is needed for GP training
+   */
+  virtual void preNeedSample() override;
+
+  /**
+   * Based on the computations in preNeedSample, the decision to get more data is passed and results
+   * from the GP fills @param val
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:35]: Document the input output arguments please.

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+   *
+   * @return bool Whether a full order model evaluation is required
+   */
+  virtual bool facilitateDecision() override;
+
+private:
+  /// Store all the outputs used for training from the LF model
+  const std::vector<Real> & _outputs_lf;
+
+  /// Store all the outputs used for training from the LF model
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:54]: The comment of this guy is the same as the one above

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  std::vector<Real> _outputs_lf_batch;
+
+  /// Communicator that was split based on samples that have rows
+  libMesh::Parallel::Communicator _local_comm;
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:58]: This communicator is already in the sampler. I don't think splitting the communicator this many times is reasonable. You can write an `get` function to the sampler's communicator.

File: modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+{
+  for (dof_id_type i = 0; i < _inputs.size(); ++i)
+  {
+    _gp_mean[i] = _gp_eval.evaluate(_inputs[i], _gp_std[i]);
+    _flag_sample[i] = !learningFunction(_outputs_lf_batch[i] + _gp_mean[i], _gp_std[i]);
+  }
+
+  for (const auto & fs : _flag_sample)
+    if (!fs)
+      return false;
```
Comment (grmnptr) [inline modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C:47]: Is there a reason why flag_sample is negated twice? 
```
_flag_sample[i] = !learningFunction(_outputs_lf_batch[i] + _gp_mean[i], _gp_std[i]);
```
and
```
if (!fs)
      return false;
```
Is it because something outside of this object needs it with the given sign?

File: modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  return true;
+}
+
+void
+BiFidelityActiveLearningGPDecision::preNeedSample()
+{
+  _outputs_lf_batch = _outputs_lf;
```
Comment (grmnptr) [inline modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C:54]: This is just a comment for future PRs. I think we should move away from using member variables for almost everything and try to design function which interface with each other using input-outputs. It is much harder to keep track of in terms of who is modifying it.

File: modules/stochastic_tools/test/tests/reporters/BFActiveLearning/gold/multiple_proc_multiple_row_ufunction.json [hunk 8]
```diff
@@ -0,0 +1,1102 @@
+{
```
Comment (grmnptr) [inline modules/stochastic_tools/test/tests/reporters/BFActiveLearning/gold/multiple_proc_multiple_row_ufunction.json:1]: For future tests. Keep the `json` output at a reasonable level. It can get really verbose really fast. You might not need to run 20 time steps to verify that your model runs and gives the expected results.

File: modules/stochastic_tools/doc/content/source/reporters/BiFidelityActiveLearningGPDecision.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# BFActiveLearningGPDecision
```
Comment (grmnptr) [inline modules/stochastic_tools/doc/content/source/reporters/BiFidelityActiveLearningGPDecision.md:1]: Could you some more information here about the object?

File: modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h [hunk 10]
```diff
@@ -39,32 +39,32 @@ class ActiveLearningGPDecision : public ActiveLearningReporterTempl<Real>,
                           dof_id_type global_ind,
                           Real & val) override;
 
-private:
   /**
-   * This evaluates the active learning acquisition function and returns bool
-   * that indicates whether the GP model failed.
+   * This makes decisions whether to call the full model or not based on
```
Comment (grmnptr) [inline modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h:43]:
```suggestion
   * Make decisions whether to call the full model or not based on
```

**Review by grmnptr — CHANGES_REQUESTED** [review]
Nice! Just a few comments.

File: modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h (hunk 1 above)
Comment (somu15) [reply modules/stochastic_tools/include/reporters/ActiveLearningGPDecision.h:83]: The derived `BiFidelityActiveLearningGPDecision` class could access the private variables.

File: modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C (hunk 6 above)
Comment (somu15) [reply modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C:47]: I just followed the logic from the base class. Looks like the even if the `!` are removed, we should get same behavior.

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h (hunk 5 above)
Comment (somu15) [reply modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:58]: Is there an example I can follow?

File: modules/stochastic_tools/test/tests/reporters/BFActiveLearning/gold/multiple_proc_multiple_row_ufunction.json (hunk 8 above)
Comment (somu15) [reply modules/stochastic_tools/test/tests/reporters/BFActiveLearning/gold/multiple_proc_multiple_row_ufunction.json:1]: I have reduced the number of samples to keep the json smaller.

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h (hunk 4 above)
Comment (somu15) [reply modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:60]: Added the input and output descriptions now.

File: modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h (hunk 3 above)
Comment (somu15) [reply modules/stochastic_tools/include/reporters/BiFidelityActiveLearningGPDecision.h:35]: Done

File: modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BiFidelityActiveLearningGPDecision.h"
+#include "Sampler.h"
+
+registerMooseObjectAliased("StochasticToolsApp",
```
Comment (grmnptr) [inline modules/stochastic_tools/src/reporters/BiFidelityActiveLearningGPDecision.C:13]: Could you please change these not to use aliases?

**Review by grmnptr — CHANGES_REQUESTED** [review]
One little thing and it is good to go!

**Review by grmnptr — APPROVED** [review]
Looks good to me! Thanks @somu15 for the nice contribution!

**Review by zachmprince — APPROVED** [review]
Great work Som!
